### DISCUSSION TITLE
AuthApiError: invalid request: both auth code and code verifier should be non-empty

### DISCUSSION THREAD

**Discussion by kafle1** [supabase/supabase#22371] [Questions]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

Error: AuthApiError: invalid request: both auth code and code verifier should be non-empty

Checked everything on the internet, all the old issues, posted on supabase discord still nothing seems to work for me.
I am trying to implement authentication with supabase and sveltekit. Everything works fine locally, but when I deploy it I get the error. Error occurs when I am redirected to /auth/callback (from email link while signup, from google in oauth). `+server.ts` file is there to handle callback and i get code in the URL which is then extracted and passed to exchangeCodeForSession() function coming from supabase auth.
Am following exactly that's mentioned in this doc (not the UI part): https://supabase.com/docs/guides/getting-started/tutorials/with-sveltekit#building-the-app

The error is being thrown from here:
```ts
const { data } = await supabase.auth.exchangeCodeForSession(code);
```
## To Reproduce
When button is clicked this is hit:
```ts
const { data } = await supabase.auth.signInWithOAuth({
			provider: provider as Provider,
			options: {
				redirectTo: `${url.origin}/auth/callback?provider=${provider}`
			}
		});
```
After auth is completed its redirected to this:
```ts
const {
url,
locals: { supabase }
} = event;
const code = url.searchParams.get('code') as string;
const provider = url.searchParams.get('provider') as string;

      if (!code) {
	      error(400, 'No code provided');
      }
      
      const { data } = await supabase.auth.exchangeCodeForSession(code);
```

## Expected behavior

Login without any error.

## System information

- OS: Zorin 17.1
- Version of supabase-js: ^2.39.8
- Version of Node.js: 20.12.0

"@supabase/ssr": "^0.1.0",
"@supabase/supabase-js": "^2.39.8",

**Comment by encima** [comment]
If you are experiencing this on a hosted project, it may be easier to open a support ticket so that the team can access logs and config to help debug

**Comment by kafle1** [comment]
Created post on discord server too but got no response :(
https://discord.com/channels/839993398554656828/1222519673782927502

this is the log:
```ts
AuthApiError: invalid request: both auth code and code verifier should be non-empty
0|app      |     at handleError (/root/FunApps.ai/node_modules/@supabase/auth-js/dist/main/lib/fetch.js:63:11)
0|app      |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
0|app      |     at async _handleRequest (/root/FunApps.ai/node_modules/@supabase/auth-js/dist/main/lib/fetch.js:111:9)
0|app      |     at async _request (/root/FunApps.ai/node_modules/@supabase/auth-js/dist/main/lib/fetch.js:89:18)
0|app      |     at async SupabaseAuthClient._exchangeCodeForSession (/root/FunApps.ai/node_modules/@supabase/auth-js/dist/main/GoTrueClient.js:409:33)
0|app      |     at async /root/FunApps.ai/node_modules/@supabase/auth-js/dist/main/GoTrueClient.js:736:28 {
0|app      |   __isAuthError: true,
0|app      |   status: 400,
0|app      |   code: undefined
0|app      | }
```

**Comment by GaryAustin1** [comment]
Did you open a support request?
Discord is not official support and is mainly user helping user.

**Reply by kafle1** [reply]
@GaryAustin1  Support ticket was opened but they responded with the same other issue which is not working for me.

```
Thank you for contacting Supabase support. I recommend checking this issue here for more details about it:

https://github.com/supabase/auth-helpers/issues/587


We also have this guide here:

https://supabase.com/docs/guides/auth/server-side/oauth-with-pkce-flow-for-ssr
```

**Reply by GaryAustin1** [reply]
Yeah, my response was just to your Discord comment and the SB employee saying to contact support.  Mainly to make clear discord is user helping user forum (I'm a user).
You have two SB devs commenting here now, so no need to flag me.  Good luck.

**Comment by charislam** [comment]
Hey @kafle1! You don't need that redirect in your `signInWithOAuth`. (Signing in with OAuth is different from signing in with password in that your redirect URL is already pre-configured, when you set up your OAuth Provider. That redirect option is useful if you want to direct to a page in your application (say your dashboard) _after_ the signin, not for redirecting to the callback route, which is automatic.)

Try this instead:

```
const { data } = await supabase.auth.signInWithOAuth({
			provider: provider as Provider,
		});
```

**Reply by kafle1** [reply]
@charislam 
The reason behind calling the `/auth/callback` route after login is that I want to get session from the code that I get after getting redirected from OAuth provider (google) and add that user's details to profile table, save avatar, create the user's default plan and other business logics which are inside my `GET` request handler in `+server.ts` file of `/auth/callback` file. After I complete my post login business logics I will redirect to user's dashboard

**Reply by charislam** [reply]
Unfortunately you won't be able to do that from an OAuth callback, because the OAuth callback is run on Supabase's servers, not on your own.

For the use case you're describing, I suggest either:

- ~~Using the `redirectTo` as you do now, but taking out the `exchangeCodeForSession` (you don' t need this). Then do your business logic there.~~ (See error correction below.) (I think this will work, haven't double-checked, it is possible you may run into a situation where you need to manually set the cookie header.)
- Using a [Postgres trigger](https://supabase.com/docs/guides/database/postgres/triggers)

**Reply by charislam** [reply]
Sorry, error correction. For the first case, it should be:

- Put the logic in your `auth/callback` route, but you don't need to set `redirectTo`.

**Reply by kafle1** [reply]
Found a temporary fix by setting this
```ts
cookieOptions: {
	secure: false
},
```
but I guess this is not the solution

**Reply by charislam** [reply]
Hmm yeah that's definitely not the solution. Though I think I may also have been wrong about the `redirectTo` param, sorry about that! Should've double-checked before responding.

I did build an example that worked on a hosted site (Vercel + Supabase hosted) here: https://github.com/charislam/svelte-example/tree/oauth-test

Maybe take a look and see how that's different from what you have?

**Reply by kafle1** [reply]
@charislam  Copied your code but still the same, am I missing something during deployment? I guess its related to cookies now

**Reply by charislam** [reply]
Hmm is your callback route getting hit at all? (if you put a console log at the top of the handler, do you see the log?)

**Reply by kafle1** [reply]
Yes its hitting and the error is the same coming from supabase

**Reply by charislam** [reply]
You said this works locally but not when deployed? Where are you deploying?

**Reply by kafle1** [reply]
Tried it in firebase, cloudflare (pages, workers), and also in virtual machine, same for all
http://172.233.136.63/

**Reply by kafle1** [reply]
and am not getting logs here too, 
```ts
if (evt === 'SIGNED_IN' || evt === 'TOKEN_REFRESHED' || evt === 'USER_UPDATED') {
		console.log('event', evt);
	} else if (evt === 'SIGNED_OUT') {
		console.log('event', evt);
	}
```

**Reply by TechWithTy** [reply]
This worked for me
